### PR TITLE
Fix #5257: Support auto generic-tupling of parameters

### LINKED ISSUE #5257: Auto tupling of function parameters for generic tuples
As a generalization of `Tuple2[Int, Int]`, `Int *: Int *: Unit` should also allow autotupling.

```scala
val h2: Int *: Int *: Unit => Int = (x, y) => x + y
```
it currently fails with

```
4 |  val h2: Int *: Int *: Unit => Int = (x, y) => x + y
  |                                      ^^^^^^^^^^^^^^^
  |                                   wrong number of parameters, expected: 1
```

### PR DESCRIPTION
**Pull request by nicolasstucki**
Additions:
* Recognise `*:` as an auto-tupleable type for function parameters
* Desugar auto-tupled function parameters for a `*:` tuples using `apply(N)` method instead of `_N`s

### CODE REVIEW

**Comment by nicolasstucki** [comment]
Rebased and added regression test

**Comment by nicolasstucki** [comment]
Rebased

File: compiler/src/dotty/tools/dotc/typer/Applications.scala [hunk 1]
```diff
@@ -59,8 +59,17 @@ object Applications {
     extractorMemberType(tp, nme.get, errorPos).exists
 
   def productSelectorTypes(tp: Type, errorPos: Position = NoPosition)(implicit ctx: Context): List[Type] = {
-    val sels = for (n <- Iterator.from(0)) yield extractorMemberType(tp, nme.selectorName(n), errorPos)
-    sels.takeWhile(_.exists).toList
+    def tupleSelectors(n: Int, tp: Type): List[Type] = {
+      val sel = extractorMemberType(tp, nme.selectorName(n), errorPos)
+      if (sel.exists) sel :: tupleSelectors(n + 1, tp) else Nil
```
Comment (OlivierBlanvillain) [inline compiler/src/dotty/tools/dotc/typer/Applications.scala]: Maybe add a comment to explain why sel can be NoType here

File: compiler/test/dotc/run-test-pickling.blacklist [hunk 2]
```diff
@@ -23,6 +23,7 @@ i4947b
 i5119
 i5119b
 i5188a
+i5257.scala
```
Comment (OlivierBlanvillain) [inline compiler/test/dotc/run-test-pickling.blacklist]: Do you know what's failing?

File: compiler/src/dotty/tools/dotc/typer/Applications.scala (hunk 1 above)
Comment (nicolasstucki) [reply compiler/src/dotty/tools/dotc/typer/Applications.scala]: added

**Pull request merged**